examples/dlint/plugins/test_plugin1.js

```diff
@@ -0,0 +1,13 @@
+export default class Plugin extends Visitor {
```

Comment:
Suggestion: add another class called `LintRule` that has static `ruleCode` method. Then add `registerRule` global method that allows to register multiple rules from a single plugin, eg.
```
registerRule(SomeRule.ruleCode(), SomeRule);
registerRule(SomeRule2.ruleCode(), SomeRule2);
```

Reply:
I might not understand your suggestion so I'd like to clarify it.

What you mean is something like as follows?

User's code:

```javascript
// some-rule.js
export default class SomeRule extends Visitor {
  static ruleCode() {
    return "some-rule";
  }

  // rule implementation...
}

// another-rule.js
export default class AnotherRule extends Visitor {
  static ruleCode() {
    return "another-rule";
  }

  // rule implementation...
}
```

then dlint executes the following JS:

```javascript
import Rule1 from "some-rule.js";
import Rule2 from "another-rule.js";

const rules = [];

function registerRule(ruleClass) {
  rules.push({
    code: ruleClass.ruleCode(),
    // https://stackoverflow.com/questions/10314338/get-name-of-object-or-class
    name: ruleClass.constructor.name,
  });
}

function runPlugins(programAst) {
  for (const rule of rules) {
    const result = new rule().collectDiagnostics(programAst);
    Deno.core.jsonOpSync("add_diagnostics", result);
  }
}

registerRule(Rule1);
registerRule(Rule2);
Deno.core.jsonOpSync("add_code", rules);
```

And then dlint calls `runPlugins` in the above JS with JSON serialized AST passed.

-----------

I have no idea if it's doable to call a JS function from Rust, but your suggestion sounds like it's doable.
I'm not familiar with deno_core so I got a little confused.

Reply:
@magurotuna yes! That looks very much what I had in mind

Reply:
```
rules.push({
    code: ruleClass.ruleCode(),
    // https://stackoverflow.com/questions/10314338/get-name-of-object-or-class
    name: ruleClass.constructor.name,
  });
```
^-- I don't think `name` is necessary here - you just want to store rule code and its constructor

Reply:
Thanks for the clarification!

One more question:

> That way diagnostic could have codes written as `pluginName/ruleCode`

What do you intend by `pluginName` here? The name of class name or other things?

Reply:
> What do you intend by pluginName here? The name of class name or other things?

Not clear right now, I guess plugin (collection of files) could register itself with any name

Reply:
Got it. Then I'll leave the current style (`@deno-lint-plugin/some-rule-code`) as it is for now. We'll discuss how to handle a collection of files as a plugin later.

The idea that has come to me was as follows.

User's each plugin file:

```javascript
// some-rule.js (not default export)
export class SomeRule extends Visitor {
  static ruleCode() {
    return "some-rule";
  }

  // rule implementation...
}

// another-rule.js (not default export)
export class AnotherRule extends Visitor {
  static ruleCode() {
    return "another-rule";
  }

  // rule implementation...
}
```

User's plugin root file:

```javascript
// my-plugin.js
// plugin's root file name is used as part of rule code in diagnostics.
// For instance, if `AnotherRule` detects an error, it outputs a diagnostic
// whose code is `my-plugin/another-rule`.

export * from "./some-rule.js";
export * from "./another-rule.js";
```

The script that dlint executes via deno_core:

```javascript
import * as plugin1 from "./my-plugin.js";

const rules = [];

function registerRule(ruleClass) {
  rules.push(ruleClass);
  Deno.core.jsonOpSync("add_code", { code: ruleClass.ruleCode()} );
}

function runPlugins(programAst) {
  for (const rule of rules) {
    const result = new rule().collectDiagnostics(programAst);
    Deno.core.jsonOpSync("add_diagnostics", result);
  }
}

for (const rule in plugin1) {
  registerRule(rule);
}
```